### PR TITLE
[serve] Default deployment name to callable name

### LINKED ISSUE #14954: Better error message for @serve.deployment decorator with no parens

### LINKED ISSUE #15134: [serve] Support @serve.deployment with no args

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Much more convenient for offline use case/local testing.

## Related issue number

Closes https://github.com/ray-project/ray/issues/14954
Closes https://github.com/ray-project/ray/issues/15134

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/tests/test_deploy.py [hunk 1]
```diff
@@ -63,11 +61,30 @@ def d(*args):
...
+    def func(*args):
+        return "hi"
+
+    @serve.deployment
+    class Class:
+        def ping(self, *args):
+            return "pong"
+
+    func._name = "func"
+    Class._name = "Class"
```
Comment (architkulkarni) [inline python/ray/serve/tests/test_deploy.py:75]: What are these lines for?

File: python/ray/serve/api.py [hunk 2]
```diff
@@ -1306,16 +1326,18 @@ class MyDeployment:
...
+            _backend_def,
+            name if name is not None else _backend_def.__name__,
             config,
             version=version,
             init_args=init_args,
             ray_actor_options=ray_actor_options)
 
-    return decorator
+    # This handles both parametrized and non-parametrized usage of the
+    # decorator. See the @serve.batch code for more details.
```
Comment (architkulkarni) [inline python/ray/serve/api.py:1339]: Thanks, this was helpful!

**Review by architkulkarni — APPROVED** [review]
Looks good to me :)

File: python/ray/serve/tests/test_deploy.py (hunk 1 above)
Comment (edoakes) [reply python/ray/serve/tests/test_deploy.py:75]: oops lol these are supposed to be an assert

**Review by simon-mo — APPROVED** [review]

**Pull request merged**